Media Store

tonyr59h opened the pull request:
TODO:
- [x] REST tests (@oguz)
- [x] Finish uploading for XMLRPC
- [x] XMLRPC tests (@tonyr59h)
- [x] Error handling (@tonyr59h)
- [x] Persistence tests (MediaStore unit tests) (@tonyr59h)
- [x] Update example app (@oguz)

tonyr59h commented:
Tests completed! I'm glad @oguzkocer got to dig into the code while helping prepare this PR, thank you 🙇 !

This one's ready for review cc @maxme @aforcier

maxme commented on fluxc/build.gradle:
```diff
@@ -59,8 +59,8 @@ dependencies {
-    compile 'com.squareup.okhttp3:okhttp:3.2.0'
-    compile 'com.squareup.okhttp3:okhttp-urlconnection:3.2.0'
+    compile 'com.squareup.okhttp3:okhttp:3.3.1'
+    compile 'com.squareup.okhttp3:okhttp-urlconnection:3.3.1'
```
Are you using something specific from okhttp 3.3.+ or you just noticed it was outdated?

tonyr59h replied:
Just noticed it was outdated, I didn't specifically update it for anything. Think it should be reverted?

maxme replied:
Nope, I was just curious ;)

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java:
```diff
@@ -0,0 +1,216 @@
+     */
+    public static List<MediaModel> mediaListFromRestResponse(MultipleMediaResponse from) {
+        List<MediaModel> media = new ArrayList<>();
+        for (int i = 0; i < from.found; ++i) {
```
I think this should be something like:
```java
for (int i = 0; i < from.media.size(); ++i) {
```

I get a OOB error on a site with a huge list of images (from.found = 1000+ while the from.media list is limited to 20)

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java:
```diff
@@ -0,0 +1,374 @@
+        }
+    }
+
+    public class OnMediaError extends OnChanged {
```
We recently changed how errors are managed https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/127 - We should drop that `OnMediaError` event and encapsulate errors in all other `OnChanged` events. 

See [OnSiteChanged](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/47417e5c9576bed50fab2f3000fc53c11f6c708e/fluxc/src/main/java/org/wordpress/android/fluxc/store/SiteStore.java#L95-L100) for example.

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java:
```diff
@@ -0,0 +1,295 @@
+                @Override
+                public void onErrorResponse(@NonNull BaseRequest.BaseNetworkError error) {
+                    AppLog.v(T.MEDIA, "VolleyError Fetching media: " + error);
+                    notifyMediaError(MediaAction.FETCH_MEDIA, media, MediaNetworkError.UNKNOWN);
```
I think we want to "interpret" BaseNetworkError here and "convert" it into a MediaNetworkError.

maxme replied:
For instance, if we get a `BaseNetworkError.NOT_FOUND`, we probably want to transform it into a `MediaNetworkError.MEDIA_NOT_FOUND`

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java:
```diff
@@ -0,0 +1,295 @@
+                        AppLog.i(T.MEDIA, "Attempted to delete media that does not exist remotely.");
+                        notifyMediaDeleted(MediaAction.DELETE_MEDIA, media);
+                    } else {
+                        notifyMediaError(MediaAction.DELETE_MEDIA, media, MediaNetworkError.UNKNOWN);
```
Same here

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java:
```diff
@@ -0,0 +1,295 @@
+                @Override
+                public void onErrorResponse(@NonNull BaseRequest.BaseNetworkError error) {
+                    AppLog.v(T.MEDIA, "VolleyError deleting media (ID=" + media.getMediaId() + "): " + error);
+                    if (error.type == BaseRequest.GenericErrorType.NOT_FOUND) {
```
(Same as here)

maxme commented:
Using a device or emulator on Android 6.0+ - and don't give the example app permission to read images:

1. Sign in a self hosted site and try to upload a picture: you get a onMediaUpload event, with a GENERIC_ERROR. In that case it would be great to have a FS_PERMISSION_DENIED_TO_READ_FILE (or PERMISSION_DENIED_BY_FILESYSTEM). So the app could either ask the user for permission or notify the user it can't access the media library.

2. Sign in a wpcom site and try to upload a picture: you never get any onMediaUpload event. We should get a onMediaUpload in all cases (error or not), and follow the same model as above.

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java:
```diff
@@ -0,0 +1,376 @@
+        }
+    }
+
+    public class OnMediaUploaded extends OnChanged<MediaError> {
```
Can we add a boolean (`completed`) or a method (`isCompleted()`) to this event? I'm not sure this is very safe to compare float as in `progress >= 1.0f`

tonyr59h replied:
Added in a5ce2ecbf5db593547043c13db526d688d4649e1

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java:
```diff
@@ -0,0 +1,386 @@
+                .post(requestBody)
+                .build();
+
+        mOkHttpClient.newCall(request).enqueue(new Callback() {
```
I know it's a PITA to use volley here, but if we use the "raw" OkHttpClient, we also have to port everything that was made for volley like the custom UserAgent and Http Auth headers like [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/BaseXMLRPCClient.java#L56-L57)

maxme replied:
BTW, we should add a huge comment somewhere to explain why we're not using volley for uploads.

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java:
```diff
@@ -0,0 +1,295 @@
+            @Override
+            public void onFailure(Call call, IOException e) {
+                AppLog.w(T.MEDIA, "media upload failed: " + e);
+                notifyMediaError(MediaAction.UPLOAD_MEDIA, null, MediaNetworkError.UNKNOWN);
```
Propagating the timeout error would be great in that case. 

For example, simulate a network disconnection: try to upload a huge image (or throttle your network), then enable airplane mode - that will generate an "UNKNOWN" error, would be great to have a TIMEOUT error instead, so the app could show the user a specific message or ask them to retry.

maxme commented:
Minor changes in the example app in 264b55e

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/module/ReleaseNetworkModule.java:
```diff
@@ -145,6 +147,25 @@ public SiteXMLRPCClient provideSiteXMLRPCClient(Dispatcher dispatcher,
+    @Singleton
+    @Provides
+    public MediaXMLRPCClient provideMediaXMLRPCClient(Dispatcher dispatcher,
+                                                      @Named("regular") RequestQueue requestQueue,
```
You must use `@Named("custom-ssl")` here and on the next line to get the certificate "manager" (`memorizingTrustManager`) for XMLRPC (ie. selfhosted sites) requests.

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/module/ReleaseNetworkModule.java:
```diff
@@ -156,7 +177,6 @@ public AccountRestClient provideAccountRestClient(Context appContext, Dispatcher
     @Provides
     public PostRestClient providePostRestClient(Context appContext, Dispatcher dispatcher,
                                                 @Named("regular") RequestQueue requestQueue,
-                                                AppSecrets appSecrets,
```
👍

tonyr59h commented:
> it would be great to have a FS_PERMISSION_DENIED_TO_READ_FILE

Added new error in 86ff78d919ef8357f97d902758713aa13210a11d and updated the example app.

Ran into a little issue: the Android lifecycle stops/starts the Activity when opening the media picker which de-registers the Activity from receiving `OnChanged` events. The upload returns with an error before `onStart` so the Activity receives no event.

Not a huge deal, just had to defer the upload request, but it got me wondering if sticky events would be a useful feature. Has it been considered?

oguzkocer commented:
@tonyr59h: How are we handling the `user doesn't have permission to see the media library` error? I can't find a relevant `MediaStore.MediaErrorType` for it.

**More context:**
I am currently working on the integration and started with the `MediaGridFragment` and changing the current `ApiHelper.SyncMediaLibraryTask` to `MediaActionBuilder.newFetchAllMediaAction(payload)` dispatch and handling the response. In the previous implementation, we have an error type `ApiHelper.ErrorType.NO_UPLOAD_FILES_CAP` which is checked to see if the user has access to see the media library in the callback. _The name is pretty misleading, but it's going away anyway_ :fire: I am just trying to find the right kind of `MediaStore.MediaErrorType` for it.

oguzkocer commented:
@tonyr59h It looks like `MediaActionBuilder.newFetchAllMediaAction(payload)` is getting `ALL` media, is there a plan for adding filters before this PR is merged in? Especially for pagination purposes, it'd be great to be able to send an offset and possibly even get it back in the event, so we know which section is fetched. What do you think?

tonyr59h commented:
> How are we handling the user doesn't have permission to see the media library error? I can't find a relevant MediaStore.MediaErrorType for it.

I think `MediaErrorType.UNAUTHORIZED` makes sense in this case. Right now it's throwing a generic error, I'll update it to recognize the unauthorized response (403) and use the proper error.

tonyr59h commented:
> is there a plan for adding filters before this PR is merged in? Especially for pagination purposes, it'd be great to be able to send an offset and possibly even get it back in the event, so we know which section is fetched. What do you think?

Thanks for pointing me to the `Filter` class in WPAndroid. Filter incoming :)

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/utils/MediaUtils.java:
```diff
@@ -0,0 +1,130 @@
+    public static boolean canReadFile(String filePath) {
+        if (filePath == null || TextUtils.isEmpty(filePath)) return false;
+        File file = new File(filePath);
+        return file.canRead();
```
I'm not sure why, but on some emulators (geny 7.0) this call `file.canRead()` returns false. The file exist, the path is correct. Nothing printed in the logs. Permission error?

tonyr59h replied:
IIRC 627f6676cd7d8761d39ef8144d49d12b3cb8f17f should have fixed that problem...I'll investigate further.

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/store/MediaStore.java:
```diff
@@ -0,0 +1,515 @@
+    }
+
+    private boolean isWellFormedForUpload(@NonNull MediaModel media) {
+        return BaseUploadRequestBody.hasRequiredData(media) == null;
```
We should log the error returned by `hasRequiredData`, can be handy to debug.

oguzkocer replied:
Added in `feature/39-media-store-small-fixes`

maxme commented:
This is still an issue:

https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/122#r77520777 - we should reuse the same User Agent, maybe something like:

```diff
--- a/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/BaseXMLRPCClient.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/BaseXMLRPCClient.java
@@ -20,7 +20,7 @@ public class BaseXMLRPCClient {
     private SiteModel mSiteModel;
     private final RequestQueue mRequestQueue;
     protected final Dispatcher mDispatcher;
-    private UserAgent mUserAgent;
+    protected UserAgent mUserAgent;
     protected OnAuthFailedListener mOnAuthFailedListener;
     private HTTPAuthManager mHTTPAuthManager;
 
diff --git a/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java b/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java
index 6217e2d..560ac93 100644
--- a/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java
+++ b/fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java
@@ -224,6 +224,7 @@ public class MediaXMLRPCClient extends BaseXMLRPCClient implements ProgressListe
         okhttp3.Request request = new okhttp3.Request.Builder()
                 .url(url)
                 .post(requestBody)
+                .addHeader("User-Agent", mUserAgent.toString())
                 .build();
 
         mOkHttpClient.newCall(request).enqueue(new Callback() {
```

For HTTP Auth, this will be a bit more tricky

maxme commented:
When trying to upload a media to a site with HTTP Auth enabled, I get the following error:

```
W/WordPress-MEDIA( 4890): error uploading media: Response{protocol=http/1.1, code=401, message=Authorization Required, url=http://[login]:[password]@bh.wpmt.co/xmlrpc.php}
D/WordPress-API( 4890): Dispatching action: MediaAction-UPLOADED_MEDIA
```

oguzkocer commented on fluxc/src/main/java/org/wordpress/android/fluxc/action/MediaAction.java:
```diff
@@ -0,0 +1,28 @@
+    // Local actions
+    @Action(payloadType = MediaListPayload.class)   UPDATE_MEDIA,
+    @Action(payloadType = MediaListPayload.class)   REMOVE_MEDIA,
+}
```
Fixed in `feature/39-media-store-small-fixes`

tonyr59h replied:
> makes the payload/action association more obvious

That's what I was going for, but it does make more sense to stick with our code style 👍

maxme commented:
On a wpcom site (I haven't tested on a self hosted yet, but that could be different):

There is something wrong when I try to upload on a slow network (throttled connection), the progress shows 99% very quickly then is stuck at 99% for 15 secs, and then I get a timeout. Using Charles I see the connection active and uploading (not stale, so the timeout at 15s is weird). 

I'm not sure why the progress is up to 99% so quickly, I wonder if we measure disk read instead of network write.

maxme commented:
Relevant logs:
```
D/WordPress-API: Dispatching action: MediaAction-UPLOAD_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 0.99
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
V/WordPress-MEDIA: Progress update on upload of null: 1.0
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
W/WordPress-MEDIA: media upload failed: java.net.SocketTimeoutException: timeout
D/WordPress-API: Dispatching action: MediaAction-UPLOADED_MEDIA
```

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/media/MediaRestClient.java:
```diff
@@ -0,0 +1,476 @@
+
+    @Override
+    public void onProgress(MediaModel media, float progress) {
+        notifyMediaProgress(media, Math.max(progress, 0.99f), null);
```
Should be `Math.min(progress, 0.99f)` here (that's why we only see 99% progress in the logs)

oguzkocer replied:
That's an unfortunate error :( Do you want me to fix that in `feature/39-media-store-small-fixes` or do you want to send a separate PR for it?

maxme replied:
Same branch `feature/39-media-store-small-fixes` will work

oguzkocer replied:
Fixed in `feature/39-media-store-small-fixes`

maxme commented on fluxc/src/main/java/org/wordpress/android/fluxc/network/xmlrpc/media/MediaXMLRPCClient.java:
```diff
@@ -0,0 +1,474 @@
+
+    @Override
+    public void onProgress(MediaModel media, float progress) {
+        notifyMediaProgress(media, Math.max(0.99f, progress), null);
```
Same here, should be `Math.min(progress, 0.99f)`.

oguzkocer commented:
As far as I can tell the permission error is the only thing holding back this PR. If I missed anything else let's list them here so we can handle them asap. Once this is merged in, we can focus on the integration part which is at least half-way done.

tonyr59h commented:
> As far as I can tell the permission error is the only thing holding back this PR.

That's my take as well. @maxme did we miss anything?

aforcier commented:
I'm getting two failing XML-RPC tests (REST tests all run fine).

##### ReleaseStack_MediaXmlRpcTest.testDeleteMediaThatExists:

I'm seeing:

```
WordPress-MEDIA: Media doesn't have required data: null
```

in the logs, but the upload/delete network requests are all going through and responding fine.

##### ReleaseStack_MediaXmlRpcTest.testPushNewMedia

I'm seeing:
```
junit.framework.AssertionFailedError: expected:<PUSHED_MEDIA> but was:<NOT_FOUND_ERROR>
```

aforcier commented:
If I use the example app to upload an image to a self-hosted site, I get a timeout:
```
W/WordPress-MEDIA: media upload failed: java.net.SocketTimeoutException: timeout
```
which results in a `GENERIC_ERROR`, but the image was actually uploaded (and sniffing the request shows a successful response arriving).

I tried this with a different (smaller) image and it worked. Here's the image that failed: https://cloudup.com/c3-NtHuD-Gn

(The tests in my above comment still fail when I use an image that uploads correctly from the example app).

aforcier commented:
This seems as good a place as any to mention it - let's decide on a naming style for test classes. With this PR, we'll have:

```
ReleaseStack_AccountTest
ReleaseStack_DiscoveryTest
ReleaseStack_MediaRestTest
ReleaseStack_MediaXmlRpcTest
ReleaseStack_PostTestWPCOM
ReleaseStack_PostTestXMLRPC
ReleaseStack_SiteTest
ReleaseStack_SiteTestWPCOM
```

and we'll also have (https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/186):

```
ReleaseStack_CommentTestWPCom
ReleaseStack_CommentTestXMLRPC
```

tonyr59h commented:
> ReleaseStack_MediaXmlRpcTest.testPushNewMedia

This test is actually irrelevant now that the auto-upload behavior of the `PUSH_MEDIA` action was removed.

tonyr59h commented:
> let's decide on a naming style for test classes

My vote is for `WPCom` semantics ☑️

aforcier commented:
> My vote is for WPCom semantics ☑️
>
> So media tests would be ReleaseStack_MediaTestWPCom and ReleaseStack_MediaTestXMLRPC.

I applied the 'WPCom & XMLRPC' naming scheme to the rest of our tests in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/196.

maxme commented:
:shipit: 🎉
